### PR TITLE
[DPE-2245][DPE-2346] Create and List backups

### PR DESCRIPTION
**Pull request by phvalguima**
This PR adds the actions `create-backup` and `list-backups`, based on [DPE-2245](https://warthogs.atlassian.net/browse/DPE-2245) and [DPE-2346](https://warthogs.atlassian.net/browse/DPE-2346).

It adds a new lib, `opensearch_backups`, composed of different classes that enumerate the potential errors with backup service, the plugin config with `OpenSearchBackupPlugin` and the `OpenSearchBackup` class, responsible for for managing the S3 relation and related actions. OpenSearchBackup receives the events from Juju and interacts with the different parts of the charm (plugin_manager, opensearch, charm, etc) to manage the backups.

The backup must happen in two stages:
1) `OpenSearchBackupPlugin`: the configuration is received completely from `S3-credentials` relation and applied to the corresponding configuration files, as any plugin; a restart is issued
2) When the user requests the first action, the second stage happens, where `OpenSearchBackup` sets the S3 repo information up in the cluster using the backup API

Configuration must be applied to `opensearch.yml`, `opensearch-keystore` and potentially, to the CA truststore if this S3 uses a self-signed certificate. 

The removal of S3 relation must be followed by the removal of any configuration in `opensearch.yml` and `opensearch-keystore`, or OpenSearch fails in the next restart as the plugin is partially configured. Removal has a simplified cycle, where (1) any API calls to disable backup are made; then (2) the configuration is removed and a restart request is issued.

~~This PR also introduces an abstract class: `Keystore`. The Keystore has default operations to be executed by any keystore, both OpenSearch's or JDK's. It also introduces a new child class: `OpenSearchCATruststore`. This class is responsible to manage the CA certificates trusted by OpenSearch, besides the certificates managed by `opensearch-security` plugin. `OpenSearchKeystore` now also inherits from this parent class.~~

~~In the case of SNAP, an additional step is needed, where the default CA truststore sits in the read-only /snap. In this case, the truststore is copied to $SNAP_COMMON folder and has its permissions updated to be `rw` only for `snap_daemon:root`. Then, the new filepath and password for the truststore are passed to the JVM via `jvm.options` configuration file.~~

~~This setup is done at install phase and only checked later on if the file persists in the same folder.~~

~~In the tarball, the `cacerts` truststore is kept in the same place: $OPENSEARCH_HOME/jdk/lib/security/cacerts.~~

[DPE-2245]: https://warthogs.atlassian.net/browse/DPE-2245?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DPE-2346]: https://warthogs.atlassian.net/browse/DPE-2346?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by phvalguima** [comment]
*TODO*
1) Add test coverage for S3 in AWS
2) CA truststore - generate password and store it in Juju secrets

**Comment by phvalguima** [comment]
Although not passing in the runners, I can see:
h-scaling: https://pastebin.canonical.com/p/NkTgbXpRr9/
tls-integration: https://pastebin.canonical.com/p/X5Tw6qnSJN/
ha-networking: https://pastebin.canonical.com/p/SdvN9zVR7J/

Does pass with more resources.

File: lib/charms/opensearch/v0/opensearch_keystore.py [hunk 1]
```diff
@@ -97,5 +122,91 @@ def _reload_keystore(self) -> None:
...
+            return self._opensearch._run_cmd(self._keytool, f"-v -list -keystore {self._keystore}")
+        except OpenSearchCmdError as e:
+            raise OpenSearchKeystoreError(str(e))
+
+    def add(self, entries: Dict[str, str]) -> None:
+        """Adds a new set of entries to the keystore."""
+        if not entries:
+            raise OpenSearchKeystoreError("Missing entries for keystore")
+        for key, value in entries.items():
+            self._add(key, value)
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_keystore.py:149]: the `_add` method content can  be put here instead of having it as a separate method.

File: lib/charms/opensearch/v0/opensearch_plugin_manager.py [hunk 2]
```diff
@@ -76,19 +88,37 @@ def plugins(self) -> List[OpenSearchPlugin]:
             plugins_list.append(new_plugin)
         return plugins_list
 
+    def get_plugin(self, plugin_class: OpenSearchPlugin) -> OpenSearchPlugin:
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_plugin_manager.py:91]: We should not expose Plugin instances. As explained in a previous comment.

File: lib/charms/opensearch/v0/opensearch_plugin_manager.py [hunk 3]
```diff
@@ -76,19 +88,37 @@ def plugins(self) -> List[OpenSearchPlugin]:
...
     def run(self) -> bool:
         """Runs a check on each plugin: install, execute config changes or remove.
 
         This method should be called at config-changed event. Returns if needed restart.
         """
+        if (
+            not self._charm.opensearch.is_started()
+            or self._charm.health.apply() != HealthColors.GREEN
+        ):
+            # If the health is not green, then raise a cluster-not-ready error
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_plugin_manager.py:116]: What is the reasoning behind this?
A cluster can be yellow when it doesn't have enough nodes to satisfy the replication constraints set on an index, or when shards are moving around.

File: lib/charms/opensearch/v0/opensearch_plugin_manager.py [hunk 4]
```diff
@@ -210,24 +248,47 @@ def _user_requested_to_enable(self, plugin: OpenSearchPlugin) -> bool:
...
+            are_configs_enabled = plugin.config().config_entries_to_add
+            stored_plugin_conf = self._opensearch_config.get_plugin(are_configs_enabled)
+            # Compare dicts
+            list_a = list(are_configs_enabled.keys())
+            list_a.sort()
+            list_b = list(stored_plugin_conf.keys())
+            list_b.sort()
+            if list_a != list_b:
+                # Keys do not match
                 return False
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_plugin_manager.py:261]: Please expressive variable names and no single letters ones :) 

```suggestion
            plugin_conf = plugin.config().config_entries_to_add
            stored_plugin_conf = self._opensearch_config.get_plugin(plugin_conf)
            return plugin_conf == stored_plugin_conf
```

File: lib/charms/opensearch/v0/opensearch_plugin_manager.py [hunk 5]
```diff
@@ -210,24 +248,47 @@ def _user_requested_to_enable(self, plugin: OpenSearchPlugin) -> bool:
...
-        return True
+            for key, val in stored_plugin_conf.items():
+                if are_configs_enabled.get(key) != val:
+                    return False
+            return True
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_plugin_manager.py:265]: I believe this is not needed anymore.

File: lib/charms/opensearch/v0/opensearch_plugin_manager.py [hunk 6]
```diff
@@ -210,24 +248,47 @@ def _user_requested_to_enable(self, plugin: OpenSearchPlugin) -> bool:
...
-        return self._opensearch.version != current_version[:num_points]
+        try:
+            # Plugins may have versioning slightly different than opensearch's:
+            version = self._opensearch.version
+            num_points = min(len(plugin.version.split(".")), len(version.split(".")))
+            return version[:num_points] != plugin.version[:num_points]
+        except OpenSearchError:
+            # A race condition may happen if plugins with relation events happen too early
+            # then, this method will be called before the cluster is ready.
+            # This exception should trigger a Plugin-specific error, which will be used
```
Comment (Mehdi-Bendriss) [inline lib/charms/opensearch/v0/opensearch_plugin_manager.py:280]: Could you explain why?

File: src/opensearch.py [hunk 7]
```diff
@@ -67,6 +70,15 @@ def install(self):
...
             raise OpenSearchInstallError()
 
+        # In opensearch snap, CA truststore should be placed in another folder than JDK's
+        # default path. That happens because JDK sits in the /snap folder and is read-only.
+        # Run this setup once, in the charm's lifecycle, as the truststore should land in
+        # the SNAP_COMMON folder.
+        ca_truststore = OpenSearchCATruststore(self._charm)
+        self.set_truststore_file()
+        # Trusstore has been created, then we need to add configuration to jvm.options
+        self._charm.opensearch_config.configure_jvm_ca_truststore(ca_truststore.get_jvm_config())
```
Comment (Mehdi-Bendriss) [inline src/opensearch.py:80]: I'm still not sure why we need the JVM trust store. We can create / use our own.

**Review by Mehdi-Bendriss — COMMENTED** [review]
Good first pass, thank you. I left some comments and questions. I believe we should review:
1. The use of the JDK trust store => we should not alter that. 
2. All the status related methods in the backup, re-evaluate the needed granularity level and merge all of the related methods into one.

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 8]
```diff
@@ -0,0 +1,641 @@
...
+        backups = {}
+        try:
+            backups = self._list_backups()
+        except OpenSearchBackupNeworkError as e:
+            event.fail(f"Failed: network error {e}")
+            return
+        except OpenSearchListBackupError():
+            event.fail("Failed: listing backups error, please check logs")
+            return
+        event.set_results({"snapshots": (json.dumps(backups)).replace("_", "-")})
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:248]: The response is a string, we need to do the `json.dumps()`

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 9]
```diff
@@ -0,0 +1,641 @@
...
+        self.charm = charm
+        # s3 relation handles the config options for s3 backups
+        self.s3_client = S3Requirer(self.charm, S3_RELATION)
+        self.framework.observe(self.charm.on[S3_RELATION].relation_departed, self.on_s3_depart)
+        self.framework.observe(self.s3_client.on.credentials_changed, self.on_s3_change)
+        self.framework.observe(self.charm.on.create_backup_action, self._on_create_backup_action)
+        self.framework.observe(self.charm.on.list_backups_action, self._on_list_backups_action)
+
+    @property
+    def _plugin(self) -> OpenSearchPlugin:
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:224]: Created `get_plugin_status` in the plugin_manager class

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 10]
```diff
@@ -0,0 +1,641 @@
...
+                        "indices": "*",
+                        "ignore_unavailable": False,
+                        "include_global_state": True,
+                        "partial": False,
+                    },
+                )
+            )
+            logger.debug(f"Create backup action request id {new_backup_id} response is: {output}")
+            # Now, wait for backup success
+            self.wait_backup_success(new_backup_id)
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:289]: I am not sure what is the correct direction here, [mysql waits for the backup](https://github.com/canonical/mysql-operator/blob/f8f49ddc9c1bbccfe2cad0eabaed3104bb970abc/lib/charms/mysql/v0/backups.py#L230) as well as psql, whereas [mongodb does not wait](https://github.com/canonical/mongodb-operator/blob/01a2aa7abf2616ba3c127a09bf1bfac6beec2860/lib/charms/mongodb/v1/mongodb_backups.py#L189C34-L189C34). Adding parameters also changes the overall behavior (opensearch would have different options than the remaining DPE charms).

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 11]
```diff
@@ -0,0 +1,641 @@
...
+            self.wait_backup_success()
+            # gets the latest value
+            # the very firs time this action is executed, _list_backups returns empty
+            # hence, the "or [0]"
+            new_backup_id = int(max(self._list_backups().keys() or [0])) + 1
+            output = self.get_service_status(
+                self._request(
+                    "PUT",
+                    f"_snapshot/{OPENSEARCH_REPOSITORY_NAME}/{new_backup_id}?wait_for_completion=true",
+                    payload={
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:279]: So, I am not sure if we want to take a snapshot of indices such as `.opensearch-security`, which contains sensitive information for that specific cluster. Otherwise, yes we can use the defaults.

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 12]
```diff
@@ -0,0 +1,641 @@
...
+            )
+            logger.debug(f"Create backup action request id {new_backup_id} response is: {output}")
+            # Now, wait for backup success
+            self.wait_backup_success(new_backup_id)
+        except OpenSearchBackupNeworkError as e:
+            event.fail(f"Failed: network error {e}")
+            return
+        except OpenSearchListBackupError():
+            event.fail("Failed: listing backups error, please check logs")
+            return
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:295]: They have different error messages back to the user

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 13]
```diff
@@ -0,0 +1,641 @@
...
+                result[elem["snapshot"]] = {
+                    field: elem[field] for field in ["version", "indices", "include_global_state"]
+                }
+                result[elem["snapshot"]]["state"] = str(self.get_snapshot_status(elem["state"]))
+        except KeyError as e:
+            logger.exception(f"_list_backups failed for output: {output} with {e}")
+            raise OpenSearchListBackupError()
+        return result
+
+    def wait_backup_success(self, backup_id=None, timeout=60) -> None:
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:344]: As raised previously, some charms wait for the backup, others don't. I will keep ours waiting until we decide how the charms should behave.

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 14]
```diff
@@ -0,0 +1,641 @@
...
+                    raise Exception()
+
+    def on_s3_change(self, event: EventBase) -> None:
+        """Calls the plugin manager config handler.
+
+        The S3 change will run over the plugin and verify if the relation is available if
+        the s3 config was completely provided. Otherwise, abandon the event and wait for
+        a next relation change.
+        """
+        if self._check_missing_s3_config_completeness():
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:365]: According to the lib, [s3 completeness](https://github.com/canonical/data-platform-libs/blob/7dfca6dc3fc774a9cad5edf19504045f3b6b929a/lib/charms/data_platform_libs/v0/s3.py#L610) is only for access-key and secret-key, whereas we also need bucket and region for sure.

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 15]
```diff
@@ -0,0 +1,641 @@
...
+            # Abandon this event, as it is still missing data
+            return
+
+        if self.s3_client.get_s3_connection_info().get("tls-ca-chain", None):
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:369]: It is set as an attribute from the relation. So, if we set it the [s3-integrator config](https://github.com/canonical/s3-integrator/blob/main/config.yaml#L31), then opensearch receives it.

However, we need to add it to the trusted store in the JVM.

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 16]
```diff
@@ -0,0 +1,641 @@
...
+            ca_truststore = OpenSearchCATruststore(self.charm)
+            # Always execute this method, we may need to replace the CA
+            try:
+                opensearch_backup_crt = "/tmp/opensearch_backup.crt"
+                with open(opensearch_backup_crt, "w") as f:
+                    f.write("\n".join(self.s3_client.get_s3_connection_info()["tls-ca-chain"]))
+                ca_truststore.add(entries={"s3": opensearch_backup_crt})
+                os.remove(opensearch_backup_crt)
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:377]: Done, added a more explicit method in Keystore object: `load_file`

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 17]
```diff
@@ -0,0 +1,641 @@
...
+                raise e
+
+        # Let run() happens: it will check if the relation is present, and if yes, return
+        # true if the install / configuration / disabling of backup has happened.
+        try:
+            if self.charm.plugin_manager.run():
+                self.charm.on[self.charm.service_manager.name].acquire_lock.emit(
+                    callback_override="_restart_opensearch"
+                )
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:389]: I've appended the post-restart logic to `_restart_opensearch`

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 18]
```diff
@@ -0,0 +1,641 @@
...
+            else:
+                self.charm.unit.status = BlockedStatus(
+                    "Unexpected error during plugin configuration, check the logs"
+                )
+                # There was an unexpected error, log it and block the unit
+                logger.error(e)
+            event.defer()
+        self.charm.unit.status = ActiveStatus()
+
+    def apply_post_restart_if_needed(self) -> None:
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:402]: Yes, that must happen after a restart and will finish the backup configuration once the plugin was installed.

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 19]
```diff
@@ -0,0 +1,641 @@
...
+           and defer this event.
+        2) If leader, run API calls to signal disable is needed
+        3) Update opensearch.yml, CA truststore and keystore
+        4) Emit a restart event
+        """
+        self.charm.unit.status = MaintenanceStatus("Disabling backup service...")
+        snapshot_status = self._check_snapshot_status()
+        if snapshot_status in [
+            BackupServiceState.SNAPSHOT_IN_PROGRESS,
+            BackupServiceState.SNAPSHOT_PARTIALLY_TAKEN,
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:456]: From the docs:
```
PARTIAL: At least one shard failed to store successfully. Can only occur if you set partial to true when taking the snapshot.
```

So, it may happen, the question is what should we do. I am removing it from this particular check, as we should be able to remove a relation correctly if we have a partially taken snapshot. I will add a warn log to raise the remove is happening but we had a partially taken snapshot.

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 20]
```diff
@@ -0,0 +1,641 @@
...
+            # 1) snapshot is either in progress or partially taken: block and defer this event
+            self.charm.unit.status = BlockedStatus(
+                f"Disabling backup not possible: {snapshot_status}"
+            )
+            event.defer()
+            return
+
+        if self.charm.unit.is_leader():
+            # 2) Run the API calls
+            self._execute_s3_depart_calls()
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:467]: I am leaving the function as we may need to add something, but I do not use it. I don't want to do anything destructive like removing the bucket in S3.

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 21]
```diff
@@ -0,0 +1,641 @@
...
+                    "type": "s3",
+                    "settings": {
+                        "bucket": bucket_name,
+                        "base_path": info.get("path", S3_REPO_BASE_PATH),
+                    },
+                },
+            )
+        )
+
+    def _check_missing_s3_config_completeness(self) -> List[str]:
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:543]: Instead of replacing this method to a boolean, I am actually adding more information in both log and status message containing what is actually missing

File: lib/charms/opensearch/v0/opensearch_distro.py [hunk 22]
```diff
@@ -73,6 +73,7 @@ def __init__(self, home: str, conf: str, data: str, logs: str, jdk: str, tmp: st
         self.certs = f"{conf}/certificates"  # must be under config
         self.certs_relative = "certificates"
         self.seed_hosts = f"{conf}/unicast_hosts.txt"
+        self.ca_truststore = f"{jdk}/lib/security/cacerts"
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_distro.py:76]: Same comment as above

File: lib/charms/opensearch/v0/opensearch_distro.py [hunk 23]
```diff
@@ -304,6 +305,11 @@ def call(
         except requests.JSONDecodeError:
             raise OpenSearchHttpError(response_body=resp.text)
 
+    @abstractmethod
+    def set_truststore_file(self) -> str:
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_distro.py:309]: the truststore comes from the snap and we need to copy it to a place where we can edit.

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 24]
```diff
@@ -0,0 +1,641 @@
...
+        self.charm.unit.status = MaintenanceStatus("Configuring backup service...")
+        #     (2) run the request; and
+        state = self._register_snapshot_repo()
+        #     (3) based on the response, set the message status
+        if state == BackupServiceState.SUCCESS:
+            self.charm.unit.status = ActiveStatus("Backup is active")
+        else:
+            self.charm.unit.status = BlockedStatus(f"Backup setup failed with {state}")
+
+    def on_s3_depart(self, event: EventBase) -> None:
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:442]: I've changed that, we should handle *_broken as it is the relation leaving.

File: lib/charms/opensearch/v0/opensearch_keystore.py (hunk 1 above)
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_keystore.py:149]: I use the _add for two different tasks: adding a value to the store, and adding a cert file

File: lib/charms/opensearch/v0/opensearch_plugin_manager.py (hunk 3 above)
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_plugin_manager.py:116]: Added the yellow as well

File: lib/charms/opensearch/v0/opensearch_plugin_manager.py (hunk 6 above)
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_plugin_manager.py:280]: I will remove this try/catch, it was likely a left over before my changes to check the cluster green/yellow health as above.

File: src/opensearch.py (hunk 7 above)
Comment (phvalguima) [reply src/opensearch.py:80]: Actually, we are using it by default whenever we spin up the JVM. What I am doing there is making a copying and adding one extra cert if we have a self-signed cert. Then, in that case, I need to tell the JVM to get its default certs from another store

File: lib/charms/opensearch/v0/opensearch_distro.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_distro.py]: Custom CAs will be addressed in another PR

**Comment by phvalguima** [comment]
Due to [an issue with protobuf, I am pinning it to 3.20 in our tests](https://matrix.to/#/!xdClnUGkurzjxqiQcN:ubuntu.com/$iL8wiSs9egQhuHx5S3Do_XzSJUlhBNf3oCKY7ACNwXU?via=ubuntu.com&via=matrix.org&via=cooperteam.net)

File: lib/charms/opensearch/v0/opensearch_plugin_manager.py [hunk 25]
```diff
@@ -222,30 +279,43 @@ def _user_requested_to_enable(self, plugin: OpenSearchPlugin) -> bool:
...
+        num_points = min(len(plugin_version), len(version))
+        return version[:num_points] != plugin_version[:num_points]
 
     def _is_plugin_relation_set(self, relation_name: str) -> bool:
         """Returns True if a relation is expected and it is set."""
         if not relation_name:
             return False
         relation = self._charm.model.get_relation(relation_name)
-        return relation is not None
+        return relation is not None and relation.units
```
Comment (juditnovak) [inline lib/charms/opensearch/v0/opensearch_plugin_manager.py:318]: Note! `relation-created`, `relation-broken`, `relation-disabled` and I'm a bit unsure about `relation-joined` (if the counter is already 1 from zero or not yet...?)

Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_plugin_manager.py:318]: Hi @juditnovak added more details in the docstring of `opensearch_plugins`, in this PR: https://github.com/canonical/opensearch-operator/pull/137/files#diff-b0054790c1d68243f1b71a3094cf7f0f989456fd7de12f17a92fd1b4c7ca30eb

File: lib/charms/data_platform_libs/v0/data_secrets.py [hunk 26]
```diff
@@ -0,0 +1,143 @@
+"""Secrets related helper classes/functions."""
```
[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_secrets.py:1]: It still does appear among the files presented for the merge.... It should be fully gone from here...

[the remark this answers is not in the document]
Comment (juditnovak) [reply lib/charms/data_platform_libs/v0/data_secrets.py:1]: I mean I see your point, but I'm kinda uncomfortable to merge "gabage" to `main` even temporarily.... When there's no particular need for that...

File: lib/charms/opensearch/v0/opensearch_plugin_manager.py (hunk 25 above)
Comment (juditnovak) [reply lib/charms/opensearch/v0/opensearch_plugin_manager.py:318]: Hmmm.... but if we recommend doesn't change the fact that adding the `units` number here will make `_is_plugin_relation_set() == False` within the `PluginManager.run()` execution.

This has consequences on the workflow if invoked in a `RelationCreated` event...

File: lib/charms/data_platform_libs/v0/data_secrets.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/data_platform_libs/v0/data_secrets.py]: @juditnovak file was removed

File: lib/charms/opensearch/v0/opensearch_plugin_manager.py (hunk 25 above)
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_plugin_manager.py]: @juditnovak implemented the `event-scope`, as we've discussed.

**Review by juditnovak — APPROVED** [review]
LGTM thought gotta confess I'm not an expert of the area...

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 27]
```diff
@@ -0,0 +1,538 @@
...
+
+    def _on_list_backups_action(self, event: ActionEvent) -> None:
+        """Returns the list of available backups to the user."""
+        if not self._check_action_can_run(event):
+            event.fail("Failed: backup service is not configured yet")
+            return
+        backups = {}
+        try:
+            backups = self._fetch_snapshots()
+            event.set_results({"snapshots": (json.dumps(backups)).replace("_", "-")})
```
Comment (delgod) [inline lib/charms/opensearch/v0/opensearch_backups.py:159]: Can we similar UX to PostgreSQL charm?
can be separate PR

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 28]
```diff
@@ -0,0 +1,538 @@
...
+        except OpenSearchError as e:
+            event.fail(
+                f"List backups action failed - {str(e)} - check the application logs for the full stack trace."
+            )
+
+    def _on_create_backup_action(self, event: ActionEvent) -> None:
+        """Creates a backup from the current cluster."""
+        if not self.charm.unit.is_leader():
```
Comment (delgod) [inline lib/charms/opensearch/v0/opensearch_backups.py:167]: just Q: why only leader? it is convenient if you can run a backup on particular (e.g. less busy) unit

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 29]
```diff
@@ -0,0 +1,538 @@
...
+
+        The get_s3_connection_info() checks if the relation is present, and if yes,
+        returns the data in it.
+
+        This method will go over the output and generate a list of missing parameters
+        that are manadatory to have be present. An empty list means everything is present.
+        """
+        missing_s3_configs = [
+            config
+            for config in ["region", "bucket", "access-key", "secret-key"]
```
Comment (delgod) [inline lib/charms/opensearch/v0/opensearch_backups.py:436]: region is not obligatory in some cases.
see examples https://charmhub.io/postgresql-k8s/docs/h-configure-s3-radosgw
can be separate PR

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 30]
```diff
@@ -0,0 +1,538 @@
...
+            event.defer()
+            return
+        self.charm.status.set(MaintenanceStatus("Disabling backup service..."))
+        snapshot_status = self._check_snapshot_status()
+        if snapshot_status in [
+            BackupServiceState.SNAPSHOT_IN_PROGRESS,
+        ]:
+            # 1) snapshot is either in progress or partially taken: block and defer this event
+            self.charm.status.set(
+                BlockedStatus(f"Disabling backup not possible: {snapshot_status}")
```
Comment (delgod) [inline lib/charms/opensearch/v0/opensearch_backups.py:341]: Maybe?
```suggestion
                MaintanceStatus(f"Disabling backup postponed till backup in progress: {snapshot_status}")
```
can be separate PR

**Review by delgod — APPROVED** [review]

File: lib/charms/opensearch/v0/opensearch_backups.py [hunk 31]
```diff
@@ -0,0 +1,538 @@
...
+                f"_snapshot/{S3_REPOSITORY}",
+                payload={
+                    "type": "s3",
+                    "settings": {
+                        "region": info.get("region"),
+                        "endpoint": info.get("endpoint"),
+                        "protocol": "http",  # TODO: roll back to https once we have the cert
+                        "bucket": info["bucket"],
+                        "base_path": info.get("path", S3_REPO_BASE_PATH),
+                    },
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply lib/charms/opensearch/v0/opensearch_backups.py:420]: yes, we can, for both

File: config.yaml [hunk 32]
```diff
@@ -33,6 +33,6 @@ options:
       Other dynamic roles are not validated.
 
   plugin_opensearch_knn:
-    default: false
+    default: true
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply config.yaml:36]: @delgod I think it was not good to leave KNN disabled by default. It is a core plugin, maintained by opensearch team directly. I see no reason to keep it as false
